### ISSUE TITLE
Phase angles in RAO animations

### ISSUE THREAD

**Issue by RubendeBruin** [mancellin/capytaine#41]
I'm trying to visualize a vessel response using the script from the cookbook. However the response seems to be out of phase with the waves. Mass and hydrostatic stiffness were generated with DAVE and are realistic.

I've imported the generated wave-forces, added mass and damping into orcaflex and calculated the response. 
- Wave forces are identical
- Added mass and damping are identical
- Hydrostatics are identical
- Natural periods (excluding added mass and damping) are identical
- The RAO amplitudes are identical
-  but the phase looks off (unfortunately orcaflex does not give me the calculated RAO phases).

The wave period (9 seconds) is above the natural periods of the barge, so I expect the barge to follow the wave surface. The orcaflex results look more realistic to me.

The wave-force rao data (as imported from capytaine) in orcaflex is set to the following:
![afbeelding](https://user-images.githubusercontent.com/34062862/92690049-826f2500-f340-11ea-9d21-3f8cc7209ef0.png)

The orcaflex animation:

![orcaflex](https://user-images.githubusercontent.com/34062862/92689191-e0027200-f33e-11ea-86cc-57b976b97f3e.gif)

Capytaine animations:

![MxMfJ9EMRc](https://user-images.githubusercontent.com/34062862/92689632-a120ec00-f33f-11ea-9549-e8152a4e6406.gif)

Reproduce:
```python
import logging
from pathlib import Path

import numpy as np
from numpy import pi

import capytaine as cpt
from capytaine.ui.vtk import Animation

logging.basicConfig(level=logging.INFO, format='%(levelname)-8s: %(message)s')

bem_solver = cpt.BEMSolver()


def generate_barge() -> cpt.FloatingBody:
    # boat = cpt.FloatingBody.from_file("boat_200.mar", file_format="mar", name="pirate ship")
    # boat.rotate_z(pi)

    length = 100
    width = 30
    draft = 4

    boat = cpt.RectangularParallelepiped(size=(length, width, 2*draft), resolution=(20,6,5))
    boat.add_all_rigid_body_dofs()
    boat.keep_immersed_part()

    K = np.array([[ 9.99999997e+01, -0.00000000e+00, -0.00000000e+00,
        -0.00000000e+00, -4.00000000e+02, -0.00000000e+00],
       [-0.00000000e+00,  1.00000000e+02, -0.00000000e+00,
         4.00000000e+02, -0.00000000e+00, -5.00000000e+03],
       [-0.00000000e+00, -0.00000000e+00,  3.01657500e+04,
        -0.00000000e+00, -0.00000000e+00, -0.00000000e+00],
       [-0.00000000e+00,  4.00000000e+02,  2.74355898e-08,
         1.90223296e+06, -0.00000000e+00, -2.00000000e+04],
       [-3.99999999e+02, -0.00000000e+00, -1.45519152e-05,
        -0.00000000e+00,  2.47779100e+07, -0.00000000e+00],
       [-0.00000000e+00, -5.00000000e+03, -0.00000000e+00,
        -2.00000000e+04, -0.00000000e+00,  2.60000000e+05]])

    I = np.array([[ 1.22999998e+04,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  1.22999998e+04,  0.00000000e+00],
       [ 0.00000000e+00,  1.22999998e+04,  0.00000000e+00,
        -1.22999998e+04,  0.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  1.22999998e+04,
         0.00000000e+00,  0.00000000e+00,  0.00000000e+00],
       [ 0.00000000e+00, -1.22999998e+04,  0.00000000e+00,
         1.24229998e+06,  0.00000000e+00,  0.00000000e+00],
       [ 1.22999998e+04,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  1.96922997e+07,  0.00000000e+00],
       [ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00,  0.00000000e+00,  1.96799997e+07]])

    # The computation of the RAO requires the values of the inertia matrix and the hydrostatic stiffness matrix.
    # They need to be computed independently.

    boat.mass = boat.add_dofs_labels_to_matrix(1000*I)

    boat.hydrostatic_stiffness = boat.add_dofs_labels_to_matrix(1000*K)
    return boat


def setup_animation(body, fs, omega, wave_amplitude, wave_direction) -> Animation:
    # SOLVE BEM PROBLEMS
    problems = [cpt.RadiationProblem(omega=omega, body=body, radiating_dof=dof) for dof in body.dofs]
    problems += [cpt.DiffractionProblem(omega=omega, body=body, wave_direction=wave_direction)]
    results = [bem_solver.solve(problem) for problem in problems]
    *radiation_results, diffraction_result = results
    dataset = cpt.assemble_dataset(results)

    # COMPUTE RAO
    dataset['RAO'] = cpt.post_pro.rao(dataset, wave_direction=wave_direction)

    # Compute the motion of each face of the mesh for the animation
    rao_faces_motion = sum(dataset['RAO'].sel(omega=omega, radiating_dof=dof).data * body.full_body.dofs[dof] for dof in body.dofs)

    # COMPUTE FREE SURFACE ELEVATION
    # Compute the diffracted wave pattern
    incoming_waves_elevation = fs.incoming_waves(diffraction_result)


    # SET UP ANIMATION
    animation = Animation(loop_duration=2*pi/omega)
    animation.add_body(body.full_body, faces_motion=wave_amplitude*rao_faces_motion)
    animation.add_free_surface(fs, wave_amplitude * (incoming_waves_elevation))
    return animation


if __name__ == '__main__':
    body = generate_barge()
    fs = cpt.FreeSurface(x_range=(-200, 200), y_range=(-100, 100), nx=50, ny=50)

    wave_direction = -0.5*pi
    omega = 2*pi / 9

    anim = setup_animation(body, fs, omega=omega, wave_amplitude=2, wave_direction=wave_direction)
    anim.run(camera_position=(-60, -60, 90), resolution=(800, 600))

    filename = f"{body.name}__omega_{omega:.2f}__beta_{wave_direction:.2f}.ogv"
    filepath = Path.cwd() / filename
    # anim.save(str(filepath), camera_position=(60, 60, 90), resolution=(800, 600))

```

**Comment by mancellin** [comment]
It could be a bug. I remember having some issues with phase on this kind of animations and tediously trying to check the phase conventions. I could have gotten it wrong.

**Comment by RubendeBruin** [comment]
@mancellin 

With a bit of trial and error I got the following movie which matches the orcaflex results quite well:

![sHTBjSBIic](https://user-images.githubusercontent.com/34062862/93135387-89988900-f6da-11ea-9ab4-5f843b85b01d.gif)

What I did was:
1. The combination of heave/roll did not seem to make sense, so I made the waves come from the opposite direction. I implemented this by specifying "wamit" convention in FreeSurface->incoming_waves: 

```python
        return (1j * problem.omega / problem.g
                * airy_waves_potential(self.mesh.faces_centers, problem, "wamit"))
```

2. Change the phase angle definition of the excitation force:
- swapped lead/lag
- changed phase 180 degrees

rao.py -> rao

```python
re = np.real(excitation)
im = np.imag(excitation)
excitation = -re + 1j * im
```

The result look convincing, but does it make any sense...???

**Comment by mancellin** [comment]
Good news: if both results can be reconciled with only minus and complex conjugation, it means it is probably just a matter of conventions.

We need to figure out the correct solution. Are you sure the Orcaflex animation is the right behavior?

A possible pitfall: Capytaine defines heave as (0, 0, 1), while some others (Wamit?) uses (0, 0, -1).

**Comment by RubendeBruin** [comment]
With (0,0,1) being UP or DOWN ?

**Comment by mancellin** [comment]
(0, 0, 1)  is up

**Comment by RubendeBruin** [comment]
I have received an orcaflex model where the hydrodyanmics are calculated by orcawave which we can use as a reference case for this. Will try to repeat the analysis with capytaine to reproduce the results.
Is there by any change a reader for wamit panel model files?

**Comment by marcogambarini** [comment]
Hi, I have some doubts about the phase of RAOs as well, while comparing with other reference solutions. 
If I understand correctly, the time dependence of all quantities is assumed to be exp(-iwt). Then the damping contribution in the equation of motion, and in the definition of impedance, should be `-1j*omega*dataset['radiation_damping']`. Instead, in the code the computation of impedance is
`A = (-omega**2*(dataset['mass'] + dataset['added_mass'])
         + 1j*omega*dataset['radiation_damping']
         + dataset['hydrostatic_stiffness'])`
Could this be the issue?

**Comment by mancellin** [comment]
Yes, there might be an issue there. Could you make some comparisons on a reference problem to be sure?

**Comment by marcogambarini** [comment]
Yes, I'm working on it, I will post here when I have something to show!
Also, in radiation problems, it looks like the boundary condition applied to the body is unit velocity in the relevant degree of freedom. So, the potential for unit amplitude oscillations (the ones to be multiplied by the RAO) should be obtained multiplying the solution by (-1j*omega) or (1j*omega), depending on conventions. Is this correct?

Thank you!

**Comment by marcogambarini** [comment]
I compared the results with B. Child's PhD thesis on wave energy converters (you can find it here [https://era.ed.ac.uk/handle/1842/5792](url)). I used the same geometry parameters; unfortunately he always considers also the added stiffness and damping of the electrical part. Anyway, I think it shows that the modified solution is close to the ones in the thesis, while the original one is not. Let me know what you think about this. (The plots I attached are at page 101).

Marco
![child](https://user-images.githubusercontent.com/71728752/125932483-b9479e53-17ed-4919-96b6-837666bc9d78.png)
![modif](https://user-images.githubusercontent.com/71728752/125932486-7bf39445-5802-4dae-99fb-a6c924dc1530.png)
![orig](https://user-images.githubusercontent.com/71728752/125932488-3358866e-f31f-483d-94d7-bb7943a7acbf.png)

**Comment by mancellin** [comment]
Note that I haven't thought in frequency domain for a while, so I'm not fully confident in my comments below.

> So, the potential for unit amplitude oscillations (the ones to be multiplied by the RAO) should be obtained multiplying the solution by (-1jomega) or (1jomega), depending on conventions. Is this correct?

Isn't it the other way around (a unit velocity corresponds to a 1/±jω amplitude)?

> Anyway, I think it shows that the modified solution is close to the ones in the thesis, while the original one is not. Let me know what you think about this

Indeed, it looks convincing.
Shouldn't the phase be opposite when using e^{-jωt} convention instead of e^{jωt}? Or did you already correct for that?

**Comment by ryancoe** [comment]
at low frequencies the phase angle should go to zero (i.e., the body is a wave follower)

**Comment by marcogambarini** [comment]
> Isn't it the other way around (a unit velocity corresponds to a 1/±jω amplitude)?

Yes, that's right. So, if the radiation problems are solved for unit velocity, to recover the free surface for unit amplitude one should multiply the radiation results by ±jω before multiplying them by the RAO.

> Shouldn't the phase be opposite when using e^{-jωt} convention instead of e^{jωt}?

It is: the modified response has increasing phase, while the original one has decreasing phase. This was obtained by just changing the sign of the radiation damping term, without further corrections.

**Comment by marcogambarini** [comment]
I was able to do some comparisons with Wamit on this test case. I will post them here in the next few days.

**Comment by mancellin** [comment]
>> Shouldn't the phase be opposite when using e^{-jωt} convention instead of e^{jωt}?
>
> It is: the modified response has increasing phase, while the original one has decreasing phase. This was obtained by just changing the sign of the radiation damping term, without further corrections.

I meant: does B. Child's code (WAMIT?) uses the same time convention as us? If not, shouldn't we expect Capytaine to return a decreasing phase to match their result?
Note also that there could be a different convention for the sign of the "heave" direction (see messages to @RubendeBruin above).

> I was able to do some comparisons with Wamit on this test case. I will post them here in the next few days.

Great! I guess the best things to do is to double-check each component (excitation force, added mass, ...) and not only the RAO.

**Comment by marcogambarini** [comment]
> does B. Child's code (WAMIT?) uses the same time convention as us?

It uses the convention e^{-jωt}, which I understood from the documentation to be the same as Capytaine's. The code is based on eigenfuction expansions.

> I guess the best things to do is to double-check each component (excitation force, added mass, ...)

For now I have some plots of free surface amplitude, but I can extract the other data as well. They were generated by correcting the radiation elevation
`rao_radiation_elevation = rao_radiation_elevation * (-1j * omega) `
and then summing incident and diffracted components (in addition to correcting the phase of the RAO as mentioned above).

![abs_1_0](https://user-images.githubusercontent.com/71728752/127470969-f7711638-4bcd-474f-b2c4-d2b08eddf4e9.jpg)
![abs_1_5](https://user-images.githubusercontent.com/71728752/127470973-aaf914bd-d455-4210-8913-467d6946c3e3.jpg)
![abs_2_0](https://user-images.githubusercontent.com/71728752/127470977-64b034a3-e539-4615-98d0-671e508bbb35.jpg)
![abs_2_5](https://user-images.githubusercontent.com/71728752/127470981-1efd4a91-d780-40f3-a735-a7879f98bac4.jpg)

**Comment by mancellin** [comment]
> It uses the convention e^{-jωt}, which I understood from the documentation to be the same as Capytaine's. The code is based on eigenfuction expansions.

Ok, alright.

> For now I have some plots of free surface amplitude, but I can extract the other data as well

Maybe it would be even better to plot the elevation somewhere as a function of time, since the issue is the phase with respect to time. But it is already quite convincing.

So, according to you, there are two issues:
* wrong sign of the dissipation term in the impedance matrix,
* missing multiplication by (-jω) because the BEM resolution is done for a unit velocity but the RAO is defined for a unit amplitude

@RubendeBruin Does this match the corrections you had to make to retrieve Orcaflex results (cf. messages earlier in this thread)?

**Comment by marcogambarini** [comment]
> So, according to you, there are two issues:
> 
>     * wrong sign of the dissipation term in the impedance matrix,
> 
>     * missing multiplication by (-jω) because the BEM resolution is done for a unit velocity but the RAO is defined for a unit amplitude
> 

Yes, these are the issues I think.

**Comment by RubendeBruin** [comment]
@mancellin , I got lost in the maths and terminology above. Are we still talking about the animation example?

I can not find documentation about what the output of capytaine is supposed to be in terms of phase angles. Are you suggesting to change the output or change the definition/documentation?

**Comment by mancellin** [comment]
Yes, it is about the phase of the RAO computed by Capytaine (although animating the RAO might still introduce more issues...).

I guess we need both: a careful writing on the phase convention in the documentation and a fix of the code since there seems to be something inconsistent there. I'll work on that when I have some time.

**Comment by marcogambarini** [comment]
I would be happy to contribute if that's ok.

**Comment by mancellin** [comment]
@marcogambarini Sure! You're welcome to submit PRs to extend the documentation or fix the code.

**Comment by marcogambarini** [comment]
#92

**Comment by marcogambarini** [comment]
OK, I have submitted a PR for some additions to the documentation. If the theory is correct, I may then fix the sign in the computation of impedance and add the term -j*omega in the computation of the free surface accordingly.

**Comment by mancellin** [comment]
Thank you for completing the theoretical documentation, it was really needed.
Feel free to submit a PR to fix the code.

To debug this kind of thing, something else we'd need is a better 1:1 mapping between the theory and the code.
When I copied them from Nemoh, I remember concluding that lines like 
https://github.com/mancellin/capytaine/blob/de208c89ddbe5592253128166c32436cfcd8c68a/capytaine/bem/problems_and_results.py#L325-L326 and https://github.com/mancellin/capytaine/blob/de208c89ddbe5592253128166c32436cfcd8c68a/capytaine/bem/problems_and_results.py#L346-L351 make sense, but I don't remember why and they can be pretty confusing...

**Comment by marcogambarini** [comment]
>  something else we'd need is a better 1:1 mapping between the theory and the code.

Do you mean adding comments or modifying the code so it fits directly to the theory (or something else)?

**Comment by mancellin** [comment]
> > something else we'd need is a better 1:1 mapping between the theory and the code.
> 
> Do you mean adding comments or modifying the code so it fits directly to the theory (or something else)?

Usually, adding a comment to explain is good, but rewriting the code in a way that is self-explanatory is even better.
Here, it might mean changing the amplitude/velocity conventions (as discussed in #92), so it would need to be done carefully...

**Comment by mancellin** [comment]
I believe this issue has been fixed by now.
